### PR TITLE
Fix bidding

### LINKED ISSUE #1692: Overabundant balance check for bid_collateral_operation
**Bug Description**
https://github.com/bitshares/bitshares-core/blob/71641eb27cfd99c583baf723a394152ae49dd47f/libraries/chain/market_evaluator.cpp#L394-L399

* If a bidder is creating a new bid, this check is fine;
* if a bidder is updating an old bid, this check doesn't make sense but only annoying, actually we should check `balance + old_bid_collateral >= new_bid_collateral`, because the old bid will be cancelled so collateral will be released then can be used in the new bid. Additionally, balance will be checked when calling `adjust_balance()` later, so the check here is not necessary.

Fixing this requires a consensus upgrade.

**Impacts**
Describe which portion(s) of BitShares Core may be impacted by this bug. Please tick at least one box.
- [ ] API (the application programming interface)
- [ ] Build (the build process or something prior to compiled code)
- [ ] CLI (the command line wallet)
- [ ] Deployment (the deployment process after building such as Docker, Travis, etc.)
- [x] DEX (the Decentralized EXchange, market engine, etc.)
- [ ] P2P (the peer-to-peer network for transaction/block propagation)
- [ ] Performance (system or user efficiency, etc.)
- [x] Protocol (the blockchain logic, consensus, validation, etc.)
- [ ] Security (the security of system or user data, etc.)
- [x] UX (the User Experience)
- [ ] Other (please add below)

**Steps To Reproduce**
Steps to reproduce the behavior (example outlined below):
1. globally settle asset USD;
2. fund account `test` with 10000 BTS;
3. bid with 9000 BTS;
4. try to update the bid to 8000 BTS, fail.

**Expected Behavior**
Able to update bid when `balance + old_bid_collateral >= new_bid_collateral`.

**Screenshots (optional)**
If applicable, add screenshots to help explain process flow and behavior.

**Host Environment**
Please provide details about the host environment. Much of this information can be found running: `witness_node --version`. 
 - Host OS:             [e.g. Ubuntu 18.04 LTS]
 - Host Physical RAM    [e.g. 4GB]
 - BitShares Version:   [e.g. 2.0.180425]
 - OpenSSL Version:     [e.g. 1.1.0g]
 - Boost Version:       [e.g. 1.65.1]
 
**Additional Context (optional)**

## CORE TEAM TASK LIST
- [ ] Evaluate / Prioritize Bug Report
- [ ] Refine User Stories / Requirements
- [ ] Define Test Cases
- [ ] Design / Develop Solution
- [ ] Perform QA/Testing
- [ ] Update Documentation

### PR DESCRIPTION
**Pull request by pmconrad**
Fixes #1692

### CODE REVIEW

**Comment by jmjatlanta** [comment]
Looks good. Being that the current code is more restrictive than this change, I assume we can remove the "else" after the hardfork. Do you concur?

**Comment by pmconrad** [comment]
> I assume we can remove the "else" after the hardfork

The else remains necessary when `!_bid`, i. e. this is an initial bid not an update.

File: libraries/chain/market_evaluator.cpp [hunk 1]
```diff
@@ -406,6 +399,21 @@ void_result bid_collateral_evaluator::do_evaluate(const bid_collateral_operation
...
+   {
+      if( _bid && d.head_block_time() >= HARDFORK_CORE_1692_TIME )
+      {
+         asset delta = o.additional_collateral - _bid->get_additional_collateral();
+         FC_ASSERT( d.get_balance(*_paying_account, _bitasset_data->options.short_backing_asset(d)) >= delta,
+                    "Cannot increase bid from ${oc} to ${nc} collateral when payer only has ${b}",
+                    ("oc", _bid->get_additional_collateral().amount)("nc", o.additional_collateral.amount)
+                    ("b", d.get_balance(*_paying_account, o.additional_collateral.asset_id(d)).amount) );
+      } else
+         FC_ASSERT( d.get_balance(*_paying_account, _bitasset_data->options.short_backing_asset(d)) >= o.additional_collateral,
```
Comment (jmjatlanta) [inline libraries/chain/market_evaluator.cpp]: Please fix long line.

File: tests/tests/swan_tests.cpp [hunk 2]
```diff
@@ -406,6 +406,28 @@ BOOST_AUTO_TEST_CASE( recollateralize )
...
+{ try {
+   init_standard_swan( 700 );
+
+   generate_blocks( HARDFORK_CORE_1692_TIME - 30 );
+
+   int64_t b2_balance = get_balance( borrower2(), back() );
+   bid_collateral( borrower2(), back().amount(1000), swan().amount(100) );
+   BOOST_CHECK_EQUAL( get_balance( borrower2(), back() ), b2_balance - 1000 );
+   GRAPHENE_REQUIRE_THROW( bid_collateral( borrower2(), back().amount(b2_balance), swan().amount(200) ), fc::assert_exception );
+   GRAPHENE_REQUIRE_THROW( bid_collateral( borrower2(), back().amount(b2_balance-999), swan().amount(200) ), fc::assert_exception );
```
Comment (jmjatlanta) [inline tests/tests/swan_tests.cpp]: Please fix long lines.

File: libraries/chain/market_evaluator.cpp [hunk 3]
```diff
@@ -406,6 +399,22 @@ void_result bid_collateral_evaluator::do_evaluate(const bid_collateral_operation
...
+   if( o.additional_collateral.amount > 0 )
+   {
+      if( _bid && d.head_block_time() >= HARDFORK_CORE_1692_TIME )
+      {
+         asset delta = o.additional_collateral - _bid->get_additional_collateral();
+         FC_ASSERT( d.get_balance(*_paying_account, _bitasset_data->options.short_backing_asset(d)) >= delta,
+                    "Cannot increase bid from ${oc} to ${nc} collateral when payer only has ${b}",
+                    ("oc", _bid->get_additional_collateral().amount)("nc", o.additional_collateral.amount)
+                    ("b", d.get_balance(*_paying_account, o.additional_collateral.asset_id(d)).amount) );
+      } else
```
Comment (abitmore) [inline libraries/chain/market_evaluator.cpp]: I guess this (and the hf check) can be removed after the hard fork? Maybe add a TODO comment.

Comment (pmconrad) [reply libraries/chain/market_evaluator.cpp]: Added.

**Pull request merged**